File: code_samples/section7/lesson/section7.py
"""
Section 7 - Hash Tables and Sets

This module demonstrates common hash-based patterns in Python using:
  - dict for key/value mappings (hash tables)
  - set for fast membership checks
  - collections.Counter for concise frequency counting

The examples mirror classic interview / algorithm problems:
  - Counting word frequencies
  - Finding the first duplicate element
  - Solving the Two Sum problem
"""

from collections import Counter
from typing import Iterable, Optional


def demo_hash() -> int:
    """
    Demonstrates basic dictionary (hash table) usage in Python.

    Key ideas:
      - dict.get(key, default) safely retrieves a value without raising KeyError
      - Missing keys can be treated as having value 0 when counting

    Returns:
      - The count associated with "apple" (used only for demonstration)

    Note:
      - This function mutates a local dictionary only and prints nothing.
      - The return value is used solely to show how values can be retrieved.
    """
    # Initialize an empty dictionary: key -> count
    freq: dict[str, int] = {}

    # Increment counts using get(..., 0) to handle missing keys
    freq["apple"] = freq.get("apple", 0) + 1     # apple -> 1
    freq["banana"] = freq.get("banana", 0) + 2   # banana -> 2

    # Safely read a value; returns 0 if "apple" does not exist
    n = freq.get("apple", 0)

    return n


def count_words(words: Iterable[str]) -> dict[str, int]:
    """
    Counts how many times each word appears in the input iterable.

    Args:
      words:
        Any iterable of strings (list, tuple, generator, etc.)

    Returns:
      A dictionary mapping word -> frequency.

    Complexity:
      - Time:  O(n), where n is the number of input words
      - Space: O(k), where k is the number of distinct words

    Implementation note:
      - collections.Counter is a specialized dict subclass optimized
        for frequency counting.
      - Wrapping it with dict(...) returns a plain dictionary.
    """
    return dict(Counter(words))


def first_duplicate(arr: Iterable[int]) -> Optional[int]:
    """
    Finds the first duplicate value encountered when scanning left-to-right.

    Example:
      [2, 5, 1, 2, 3, 5, 1] -> 2

    Args:
      arr:
        Iterable of integers to scan.

    Returns:
      - The first duplicated integer if found
      - None if all values are unique

    Complexity:
      - Time:  O(n) average
      - Space: O(n) worst case (when all elements are unique)

    Notes:
      - Using a set provides average O(1) membership checks.
      - Returning Optional[int] avoids ambiguous sentinel values
        like 0 or -1.
    """
    # Tracks values that have already been seen
    seen: set[int] = set()

    for x in arr:
        # If x is already in the set, it is the first duplicate encountered
        if x in seen:
            return x

        # Otherwise, record that we've seen x
        seen.add(x)

    # No duplicates found
    return None


def two_sum(nums: list[int], target: int) -> tuple[int, int]:
    """
    Solves the classic Two Sum problem.

    Finds indices (i, j) such that:
      nums[i] + nums[j] == target

    Args:
      nums:
        List of integers.
      target:
        Desired sum.

    Returns:
      - A tuple (i, j) if a valid pair is found
      - (-1, -1) if no such pair exists

    Complexity:
      - Time:  O(n) average
      - Space: O(n)

    Approach:
      - Maintain a dictionary mapping value -> index seen so far.
      - For each element x at index i:
          need = target - x
        If need has already been seen, we have a solution.

    Notes:
      - If a value appears multiple times, the dictionary stores
        the most recent index.
      - Any valid pair is acceptable for this problem.
    """
    # Map from value -> index where it was last seen
    index_of: dict[int, int] = {}

    for i, x in enumerate(nums):
        # Compute the complement needed to reach target
        need = target - x

        # If we've already seen the needed value, return indices
        if need in index_of:
            return index_of[need], i

        # Record the current value and its index
        index_of[x] = i

    # No pair found
    return -1, -1


# ============================
# Tests (top-level statements)
# ============================
#
# These statements run immediately when the module is executed.
# No explicit main() function is required.
#

print("=== Test: demo_hash ===")
demo_hash()
print("(demo_hash ran - no visible output)\n")

# ------------------------------------------------------------

print("=== Test: count_words (word frequency) ===")

# Input words:
# apple x3, banana x2, orange x1
word_list = ["apple", "banana", "apple", "orange", "banana", "apple"]
frequency = count_words(word_list)

# Print selected keys, including one that does not exist ("pear")
for key in ["apple", "banana", "orange", "pear"]:
    print(f"'{key}' -> {frequency.get(key, '(not found)')}")

print("Expected: apple=3, banana=2, orange=1, pear not found\n")

# ------------------------------------------------------------

print("=== Test: first_duplicate ===")

# Case 1: duplicates exist
arr1 = [2, 5, 1, 2, 3, 5, 1]
print("Array 1:", arr1)
print("First duplicate:", first_duplicate(arr1), "(expected 2)\n")

# Case 2: no duplicates
arr2 = [1, 2, 3, 4, 5]
print("Array 2:", arr2)
print("First duplicate:", first_duplicate(arr2), "(expected None)\n")

# ------------------------------------------------------------

print("=== Test: two_sum ===")

# Positive case: 2 + 7 = 9
numbers = [2, 7, 11, 15]
tgt = 9
print("Array:", numbers)
print("Target:", tgt)

pair1 = two_sum(numbers, tgt)
print("two_sum result indices:", pair1)
print("Expected: (0, 1) or (1, 0)\n")

# Negative test: no valid pair
print("--- Negative case (no pair) ---")
nums2 = [1, 2, 3]
target2 = 100
print("Array:", nums2)
print("Target:", target2)

pair2 = two_sum(nums2, target2)
print("two_sum result indices:", pair2)
print("Expected: (-1, -1)\n")

File: code_samples/section7/lesson/section7.js
/**
 * Demonstration of hash-based data structures in JavaScript:
 *  - Map for key/value associations (like a hash table)
 *  - Set for fast membership checks
 *
 * These examples mirror common interview / algorithm patterns:
 *  - Frequency counting
 *  - Detecting duplicates
 *  - Solving the Two Sum problem
 *
 * Notes on Map vs plain objects:
 *  - Map preserves insertion order and allows any value type as a key.
 *  - Map does not have prototype keys, so no accidental collisions like "__proto__".
 */

// ===========================================================
// Core Demonstrations
// ===========================================================

/**
 * demoHash()
 *
 * Purpose:
 *  - Shows basic usage of JavaScript's Map for counting values.
 *
 * Key idea:
 *  - map.get(key) returns undefined if the key does not exist.
 *  - Using (map.get(key) || 0) treats missing entries as 0.
 *
 * Side effects:
 *  - Mutates a local Map only; produces no visible output.
 */
function demoHash() {
    // Create an empty Map: key -> count
    const freq = new Map();

    // Increment "apple":
    //  - freq.get("apple") is undefined initially
    //  - undefined || 0 => 0
    //  - 0 + 1 => 1
    freq.set("apple", (freq.get("apple") || 0) + 1);

    // Increment "banana" by 2 using the same pattern
    freq.set("banana", (freq.get("banana") || 0) + 2);

    // Safe read with default fallback
    const n = freq.get("apple") || 0;

    // n === 1 here; value is unused, just for demonstration
}

/**
 * countWords(words)
 *
 * Purpose:
 *  - Builds a frequency map of words.
 *
 * @param {string[]} words - Array of words/tokens to count.
 * @returns {Map<string, number>} Map from word -> frequency.
 *
 * Complexity:
 *  - Time:  O(n) average
 *  - Space: O(k), where k is the number of distinct words
 */
function countWords(words) {
    // Frequency table
    const freq = new Map();

    for (const w of words) {
        // Read existing count (or 0 if missing) and increment
        freq.set(w, (freq.get(w) || 0) + 1);
    }

    return freq;
}

/**
 * firstDuplicate(arr)
 *
 * Purpose:
 *  - Returns the first value that appears more than once when scanning left-to-right.
 *
 * Example:
 *  - [2, 5, 1, 2, 3, 5, 1] -> 2
 *
 * @param {number[]} arr - Array of numbers to scan.
 * @returns {number|null} The first duplicate, or null if none exists.
 *
 * Complexity:
 *  - Time:  O(n) average
 *  - Space: O(n) worst case
 */
function firstDuplicate(arr) {
    // Set of values we've already seen
    const seen = new Set();

    for (const x of arr) {
        // If x is already in the set, we've found the first duplicate
        if (seen.has(x)) {
            return x;
        }
        // Otherwise record that we've seen x
        seen.add(x);
    }

    // No duplicates found
    return null;
}

/**
 * twoSum(nums, target)
 *
 * Purpose:
 *  - Finds indices (i, j) such that nums[i] + nums[j] == target.
 *
 * @param {number[]} nums - Input array of numbers.
 * @param {number} target - Desired sum.
 * @returns {number[]} A pair of indices [i, j], or [-1, -1] if no solution exists.
 *
 * Complexity:
 *  - Time:  O(n) average
 *  - Space: O(n)
 *
 * Approach:
 *  - Use a Map from value -> index.
 *  - For each element nums[i], compute the needed complement.
 *  - If the complement exists in the map, return the stored index and i.
 */
function twoSum(nums, target) {
    // Map: value -> index where it was last seen
    const indexOf = new Map(); // value -> index

    for (let i = 0; i < nums.length; i++) {
        const need = target - nums[i];

        // Check if we've already seen the needed value
        if (indexOf.has(need)) {
            return [indexOf.get(need), i];
        }

        // Store the current value and its index
        indexOf.set(nums[i], i);
    }

    // No valid pair found
    return [-1, -1];
}

// ===========================================================
// Helper Functions for Testing
// ===========================================================

/**
 * printMap(label, map, keys)
 *
 * Purpose:
 *  - Prints selected keys from a Map in a readable format.
 *
 * @param {string} label - Heading printed before the entries.
 * @param {Map} map - Map to inspect.
 * @param {string[]} keys - Keys to print.
 */
function printMap(label, map, keys) {
    console.log(label);

    for (const k of keys) {
        if (map.has(k)) {
            console.log(`  '${k}' -> ${map.get(k)}`);
        } else {
            console.log(`  '${k}' -> (not found)`);
        }
    }
}

/**
 * printArray(label, arr)
 *
 * Purpose:
 *  - Prints an array in JSON-style formatting.
 *
 * @param {string} label - Prefix before the array.
 * @param {any[]} arr - Array to print.
 */
function printArray(label, arr) {
    console.log(label + JSON.stringify(arr));
}

// ===========================================================
// Tests
// ===========================================================

console.log("=== Test: demoHash ===");
demoHash();
console.log("(demoHash ran - no visible output)\n");

// ------------------------------------------------------------

console.log("=== Test: countWords (word frequency) ===");

// Sample input: apple x3, banana x2, orange x1
const words = ["apple", "banana", "apple", "orange", "banana", "apple"];

// Run function under test
const freq = countWords(words);

// Print selected keys, including a missing one ("pear")
printMap("Frequencies:", freq, ["apple", "banana", "orange", "pear"]);
console.log("Expected: apple=3, banana=2, orange=1, pear not found\n");

// ------------------------------------------------------------

console.log("=== Test: firstDuplicate ===");

// Case 1: duplicates exist
const arr1 = [2, 5, 1, 2, 3, 5, 1];
printArray("Array 1: ", arr1);
console.log("First duplicate:", firstDuplicate(arr1), "(expected 2)\n");

// Case 2: no duplicates
const arr2 = [1, 2, 3, 4, 5];
printArray("Array 2: ", arr2);
console.log("First duplicate:", firstDuplicate(arr2), "(expected null)\n");

// ------------------------------------------------------------

console.log("=== Test: twoSum (hash map) ===");

// Positive case: 2 + 7 = 9
const nums = [2, 7, 11, 15];
const target = 9;

printArray("Array: ", nums);
console.log("Target:", target);

const pair1 = twoSum(nums, target);
console.log("twoSum result indices:", JSON.stringify(pair1));
console.log("Expected indices [0,1] or [1,0] (values 2 and 7)\n");

// Negative test
console.log("--- Negative case (no pair) ---");

const nums2 = [1, 2, 3];
const target2 = 100;

printArray("Array: ", nums2);
console.log("Target:", target2);

const pair2 = twoSum(nums2, target2);
console.log("twoSum result indices:", JSON.stringify(pair2));
console.log("Expected: [-1, -1]\n");
